lib/Parse/Lexer.cpp

```diff
@@ -1163,9 +1176,13 @@ unsigned Lexer::lexCharacter(const char *&CurPtr, char StopQuote,
     }
     // Move the pointer back to EOF.
     --CurPtr;
-    LLVM_FALLTHROUGH;
+    if (EmitDiagnostics)
+      diagnose(CurPtr-1, diag::lex_unterminated_string);
+    return ~1U;
   case '\n':  // String literals cannot have \n or \r in them.
   case '\r':
+    if (Modifiers & StringLiteralMultiline) // ... unless they are multiline
+      return CurPtr[-1];
```

Comment:
Try this:

```c
case '\n':
  // String literals cannot have \n in them unless they are multiline.
  if (Modifiers & StringLiteralMultiline)
    return CurPtr[-1];
  if (EmitDiagnostics)
    diagnose(CurPtr-1, diag::lex_unterminated_string);
  return ~1U;
case '\r':
  // String literals cannot have \r in them unless they are multiline, in which
  // case \r and \r\n are each normalized to \n.
  if (Modifiers & StringLiteralMultiline) {
    if (*CurPtr == '\n')
      ++CurPtr;
    return '\n';
  }
  if (EmitDiagnostics)
    diagnose(CurPtr-1, diag::lex_unterminated_string);
  return ~1U;
```

If this works, you can rip out a bunch of code you've got elsewhere.